### PR TITLE
Refactor ragged tensor processing for readbility and improved performance

### PR DESCRIPTION
**Pull request by oliverholworthy**
Refactor ragged tensor processing for readbility and improved performance

### CODE REVIEW

File: tests/unit/dataloader/test_tf_dataloader.py [hunk 1]
```diff
@@ -98,63 +98,6 @@ def test_simple_model():
     _ = model.evaluate(loader)
 
 
-def test_nested_list():
```
[the remark this answers is not in the document]
Comment (oliverholworthy) [reply tests/unit/dataloader/test_tf_dataloader.py:101]: This PR doesn't change the output representation. the output types and shapes should be the same as before. Apart from the torch version which now outputs the offsets with the last value correctly and consistently returns the scalar columns as 2-d arrays with the last dim of size 1

The reason for the removal of the test is mostly because it is now broken due to the difference between our implementation of the [pull_apart_list ](https://github.com/NVIDIA-Merlin/dataloader/blob/v0.0.4/merlin/dataloader/loader_base.py#L575) behaving in a different way between cudf/pandas. I'll try restoring these two lines https://github.com/NVIDIA-Merlin/dataloader/blob/v0.0.4/merlin/dataloader/loader_base.py#L577-L579 and see If I can restore this test.

The current implementation and what this test is checking isn't really nested list support. It currently loses the information about the nesting during the transformation. And what you get out of the dataloader cannot be turned back into the nested lists in the original data.

[the remark this answers is not in the document]
Comment (oliverholworthy) [reply tests/unit/dataloader/test_tf_dataloader.py]: Updated to restore this test `test_nested_list`, and tests are all passing. We can leave this to another PR if we want to change the behaviour. Thanks for checking about this.

[the remark this answers is not in the document]
Comment (edknv) [reply tests/unit/dataloader/test_tf_dataloader.py]: Thanks for the explanation.

File: merlin/dataloader/loader_base.py [hunk 2]
```diff
@@ -544,40 +491,22 @@ def _create_tensors(self, gdf):
...
-
-        if not offsets.empty:
-            offsets_tensor = self._to_tensor(offsets)
-            if len(offsets_tensor.shape) == 1:
-                offsets_tensor = offsets_tensor[:, None]
-            tensors.append(offsets_tensor)
-        del gdf, offsets
+                    tensors_by_name[column_name] = self._to_tensor(leaves), self._to_tensor(
+                        col_offsets
+                    )
```
Comment (edknv) [inline merlin/dataloader/loader_base.py:496]: Super nit: I think black formatting makes this harder to read (at least for me). I wonder if using parens or `tuple()`, e.g., `tensors_by_name[column_name] = (self._to_tensor(leaves), self._to_tensor(col_offsets))`, improves readability with `black`.

File: merlin/dataloader/torch.py [hunk 3]
```diff
@@ -144,11 +141,7 @@ def _pull_values_offsets(self, values_offset):
         else:
             values = values_offset.flatten()
             offsets = torch.arange(values.size()[0], device=self.device)
-        num_rows = len(offsets)
-        if HAS_GPU:
-            offsets = torch.cat([offsets, torch.cuda.LongTensor([len(values)], device=self.device)])
-        else:
-            offsets = torch.cat([offsets, torch.LongTensor([len(values)])])
```
Comment (edknv) [inline merlin/dataloader/torch.py:151]: I remember there was some issue with NVTabular and/or T4Rec in the multi-GPU case and `device=self.device` was added to line 149. What do you think about still keeping something like `if HAS_GPU: offsets = offsets.to(device=self.device)`?

**Review by edknv — COMMENTED** [review]
This is awesome! I have one more comment.

File: merlin/dataloader/torch.py (hunk 3 above)
Comment (edknv) [reply merlin/dataloader/torch.py:151]: On second thought, it was probably an issue because of `torch.cat` and `offsets` and the other tensor being on different devices, so it should be okay with the refactor.

**Review by edknv — APPROVED** [review]

File: merlin/dataloader/torch.py (hunk 3 above)
Comment (oliverholworthy) [reply merlin/dataloader/torch.py:151]: the reason for the removal of the lines doing torch.cat is because the offsets previously were missing the last value. After this change the offsets are now computed in the `_row_lengths_to_offests` method which doesn't require this logic. There's still a torch.cat in use in that method, that concatenates the cumulative sum of the row lengths with a leading zero.
